scipy/optimize/zeros.py

```diff
@@ -187,6 +187,9 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     # Multiply by 1.0 to convert to floating point.  We don't use float(x0)
     # so it still works if x0 is complex.
     p0 = 1.0 * x0
+    # check if initial guess is the solution
+    if func(p0, *args) == 0:
+        return _results_select(full_output, (p0, 0, 0, _ECONVERGED))
```

Comment:
The check outside the loop only handles the case of the starting point being a root.  Its doesn't handle an intermediate iterate being a multiple root.
Move the test inside the loop, and move the evaluation of `f` ahead of that of `fprime`.  Something like
```
        for itr in range(maxiter):
            # First evaluate f
            fval = func(p0, *args)
            funcalls += 1
            # If fval is 0, a root has been found. Terminate.
            if fval == 0:
                return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
            # Now evaluate f'
            fder = fprime(p0, *args)
            funcalls += 1
            # If fder is 0 and fval != 0, there is a problem. Warn/terminate
            if fder == 0:
                msg = "derivative was zero."
                warnings.warn(msg, RuntimeWarning)
                return _results_select(full_output, (p0, funcalls, itr + 1, _ECONVERR))
            newton_step = fval / fder
           ...

```

Reply:
I agree, in that case shall I change the title of the PR to "check for convergence using abs of residual" or something like that?

And more importantly, what tolerance should we use to check for convergence in this way? I think a test for zero identically is too strict, perhaps something more like:

```python
# If a root has been found within some tolerance, then terminate
    if abs(fval) < tol:
                return _results_select(full_output, (p0, funcalls, itr, _ECONVERGED))
```

Should we use the same tolerance as what's currently used for the Newton step or add a new argument for a separate one? Or should we just check if the guess is identically zero?

Reply:
Testing for "exactly 0" fixes the situation of landing in a multiple root.
Testing for "approximately 0" would change existing results.  It also isn't scale-invariant (replacing `f` with `const * f` leads to different outcomes when using "approximately 0".).
So "exactly 0" is the right criteria to apply.